Mode base sdk implementation (#27504)
project-chip/connectedhomeip: src/app/clusters/mode-base-server/README.md

# Mode Base and its aliases

ModeBase is a pseudo cluster. It has no cluster ID. It exists only to be derived
into other clusters.

# How to use a ModeBase derived cluster

To use a ModeBase derived cluster, you need to

-   Create a class that inherits the `ModeBase::Instance` class. For examples
    look at the `<alias name>-mode.*` files in the `src` and `include`
    directorins in `examples/all-clusters-app/all-clusters-common`.
-   For this class implement the `NumberOfModes`, `GetModeLabelByIndex`,
    `GetModeValueByIndex` `GetModeTagsByIndex` and `HandleChangeToMode` methods.
    Look at the documentation in `mode-base-server.h` for more information about
    these methods. Cluster-specific enums can be accessed from the cluster
    namespace.
-   Optionally implement the `AppInit` function.

> -   In your main file, instantiate your `ModeBase::Instance` inherited class.

Author: Yes, no need to be in main.

> -   In your main file, instantiate your `ModeBase::Instance` inherited class.

Author: Update.

# Mode Base and its aliases

Mode Base is a pseudo cluster. It has no cluster ID. It exists only to be
derived from by other clusters.

> # How to use a ModeBase derived cluster

Reviewer: ```suggestion
# How to use a Mode Base derived cluster
```

> To use a ModeBase derived cluster, you need to

Reviewer: ```suggestion
To use a Mode Base derived cluster, you need to
```

-   Create a class that inherits the `ModeBase::Delegate` class. For simple
    examples that store all their data in memory, look at the
    `<alias name>-mode.*` files in the `src` and `include` directories in
    `examples/all-clusters-app/all-clusters-common`.
-   For this class implement the `GetModeLabelByIndex`, `GetModeValueByIndex`,
    `GetModeTagsByIndex` and `HandleChangeToMode` methods. Look at the
    documentation in `mode-base-server.h` for more information about these
    methods. Cluster-specific enums can be accessed from the cluster namespace.
-   Optionally implement the `Init` function.

> -   In your translation unit, instantiate your `ModeBase::Instance` inherited

Reviewer: ```suggestion
-   In some translation unit (.c or .cpp file), instantiate your `ModeBase::Instance` inherited
```

    class. See the constructor documentation in `mode-base-server.h`.
-   Call the `.Init()` function of your instance after the root
    `Server::Init()`.
-   Alternatively, the last two steps can be done in the
    `emberAf<ClusterName>ClusterInitCallback` function.
-   Add `#define EMBER_AF_PLUGIN_MODE_BASE` to your
    `chip_device_project_config_include` file. In the examples, this file is

>     `CHIPProjectAAConfig.h`.

Reviewer: ```suggestion
    `CHIPProjectAppConfig.h`.
```

> **Note** Zap accessor functions for these clusters will not return correct

Reviewer: Is this note even needed, given that those ZAP accessor functions don't even exist because we annotated the attributes as being AttributeAccessInterface-only?  Alternately, should this note be rephrased slightly to say "do not exist" instead of "will not return correct values"?

Author: Hmm. I didn't realise that that removed those accessor functions. I will update the text

values. Use the instance's `Update...` and `Get...` functions to access the
attributes.

# How to add new derived clusters

> Once a ModeBase derived cluster has been defined in the spec, add the

Reviewer: ```suggestion
Once a Mode Base derived cluster has been defined in the spec, add the
```
